scripts/chatty_news.js

```diff
@@ -21,154 +21,39 @@ settingsLoadedEvent.addHandler(function()
-                    var head = document.getElementsByTagName('head')[0];
-                    var style = document.createElement('style');
-                    style.type = 'text/css';
-                    if (style.styleSheet){
-                        style.styleSheet.cssText = styletext;
-                    } else {
-                        style.appendChild(document.createTextNode(styletext));
-                    }
-                    head.appendChild(style);
-                }
-            },
-            sendAjax: function SendAjax(url, callbackDoc, callbackText) {
-                if (window.XMLHttpRequest) {
-                    var request = new XMLHttpRequest();
-                } else {
-                    var request = new ActiveXObject("Microsoft.XMLHTTP");
-                }
-             
-                request.open("GET", url, true);
-                request.onreadystatechange = function() {
-                    if (request.readyState == 4 && request.status == 200) {
-                        if (request.responseText) {
-                            //console.log("AjaXResponse to "+url+" doc="+callbackDoc+" text="+callbackText+" resplength="+(request.responseText.length()));
-                            if(callbackDoc!=null){
-                                if (window.ActiveXObject) {
-                                    var doc = new ActiveXObject("Microsoft.XMLDOM");
-                                    doc.async = "false";
-                                    doc.loadXML(request.responseText);
-                                } else {
-                                    var parser = new DOMParser();
-                                    //replace handles badly (nonexistant) html encoded article titles
-                                    var doc = parser.parseFromString(chattynews.html_entity_decode(request.responseText),"text/xml"); //.replace(/& /g,"&amp;")
-                                }
-                                callbackDoc(doc.documentElement);
-                            }
-                            if(callbackText!=null){
-                                callbackText(request.responseText);
-                            }
-                        }
-                    }
-                    if (request.readyState == 4 && request.status != 200) {
-                        console.log("Requested '"+url+"' bur received status "+request.status+"  text:"+request.responseText);
-                        if(callbackDoc!=null){
-                            callbackDoc(null);
-                        }
-                        if(callbackText!=null){
-                            callbackText(null);
-                        }
+                    var rssurl = "https://www.shacknews.com/feed/rss";
+                    if(document.location.href.indexOf("https")==-1){
+                        rssurl = "http://www.shacknews.com/feed/rss";
                     }
+                    xhrRequest(rssurl)
```

Comment:
That sounds great, I'll make this change tomorrow

Reply:
I'm trying to use your modified fetchSafe with XML support, but I'm getting an error back as the document fragment.  The sanitization that is being done is removing tags that aren't valid HTML tags (even though they are valid XML), however, those tags are necessary for the structure of the XML document. This is causing the DOMParser call in the fetchSafe function to fail to parse the document, and instead return a error.

The error is: 
```html
<parsererror xmlns="http://www.w3.org/1999/xhtml" style="display: block; white-space: pre; border: 2px solid #c77; padding: 0 1em 0 1em; margin: 1em; background-color: #fdd; color: black"><h3>This page contains the following errors:</h3><div style="font-family:monospace;font-size:12px">error on line 4 at column 3: Extra content at the end of the document
</div><h3>Below is a rendering of the page up to the first error.</h3></parsererror>Shacknews
```


The URL I'm requesting is https://www.shacknews.com/feed/rss .
This should look something like this:
```xml
<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0" xmlns:media="http://search.yahoo.com/mrss/" xmlns:atom="http://www.w3.org/2005/Atom">
	<channel>
		<title>Shacknews</title>
		<link>https://www.shacknews.com/</link>
		<description>Shacknews</description>
		<atom:link href="https://www.shacknews.com/feeds/rss/" rel="self" type="application/rss+xml" />

				<item>
		<guid>https://www.shacknews.com/article/112874/watch-eleagues-gears-5-multiplayer-esports-tournament-here</guid>
		<title><![CDATA[Watch ELEAGUE's Gears 5 multiplayer esports tournament here]]></title>
		<link>https://www.shacknews.com/article/112874/watch-eleagues-gears-5-multiplayer-esports-tournament-here</link>
		<pubDate>Sat, 13 Jul 2019 13:00:00 -0700</pubDate>
		<description><![CDATA[ELEAGUE is kicking off the Gears Summer Series Invitational today, so if you're looking for a first look at Gears 5 multiplayer, you've found the right place.]]></description>
		<media:thumbnail url="https://d1lss44hh2trtw.cloudfront.net/assets/article/2019/07/12/eleague-gears-5_688x387.jpg" height="525" width="295" />
		</item>
```
 . . .

What the sanitization returns is the following:
```xml
<title>Shacknews</title>
		https://www.shacknews.com/
		Shacknews
		
		https://www.shacknews.com/article/112874/watch-eleagues-gears-5-multiplayer-esports-tournament-here
		<title>&lt;![CDATA[Watch ELEAGUE's Gears 5 multiplayer esports tournament here]]&gt;</title>
		https://www.shacknews.com/article/112874/watch-eleagues-gears-5-multiplayer-esports-tournament-here
		Sat, 13 Jul 2019 13:00:00 -0700
 ```
 . . .

The XML parser thinks that the first / only element is <title>, and complains that there is data after the end of the document. Because the structure of the document is ruined by the sanitization (all tags that aren't valid html tags, such as <rss>, <channel>, <item>, etc. are being removed), it is making the XML impossible to parse.

Reply:
I'll investigate.

Reply:
Actually, I'm not sure it's possible to 'fix'.  DOMPurify doesn't support XML, as basically any tag is legal anywhere in XML.
I'm making a change to my code that should address this.  I get the XML how I previously was, but now I'm putting each value I pull out of the XML through DOMPurify before adding it to the Shack's HTML.  Not as obvious that it getting cleaned (as it isn't all done at once as the XML is fetched), but it this will prevent any script tag or other possibly dangerous html in the XML from being added to the page.

Reply:
Running each value through a sanitizer won't help matters, as it doesn't check for an XSS attack vector (which could include fetching unsafe data while DOMParser is chewing on the XML). I'm looking into what existing libraries are available that will work for this case.

Reply:
Pull the latest commits from `master`, and use the new sanitized RSS JSON object like so:

`fetchSafe(rssUrl, {}, { rssBool: true }).then(handleRSS);`

You'll get a sanitized JSON object back representing - hopefully - all the data that was contained in the original RSS document. You can simply use normal ES5/ES6 methods to iterate over this as per your needs. I think this should satisfy AMO's requirements as the RSS is never directly parsed by the browser and thus is significantly less likely to be affected by XSS attacks. On top of this I'm running the returned JSON through our `safeJSON()` helper to deep clean the object of any remaining malicious tags.